DolbyIO/Source/SdkAccessModule/Private/DeviceManagement.cpp

```diff
@@ -3,132 +3,279 @@
+			        if (IsInput(Event.utilized_direction))
 			        {
-				        if (IsInput(Event.device))
+				        if (Event.no_device)
 				        {
-					        InputDevices->OnChangedToNone();
+					        UpdateCurrentInputDeviceIndex(-1, "<None>");
 				        }
-				        if (IsOutput(Event.device))
+				        else
 				        {
-					        OutputDevices->OnChangedToNone();
+					        UpdateCurrentInputDeviceIndex(Event.device);
 				        }
 			        }
-			        else
+			        if (IsOutput(Event.utilized_direction))
 			        {
-				        if (Event.utilized_direction & dvc_device::input)
+				        if (Event.no_device)
 				        {
-					        InputDevices->OnChanged(Event.device);
+					        UpdateCurrentOutputDeviceIndex(-1, "<None>");
 				        }
-				        if (Event.utilized_direction & dvc_device::output)
+				        else
 				        {
-					        OutputDevices->OnChanged(Event.device);
+					        UpdateCurrentOutputDeviceIndex(Event.device);
 				        }
 			        }
 		        })
-		    .then([this](auto&& Event) { Events->AddEvent(MoveTemp(Event)); })
-		    .on_error(DLB_HANDLE_ASYNC_EXCEPTION);
+		    .on_error(MakeHandler(__LINE__));
 
 		DeviceManagement
 		    .add_event_handler(
-		        [this](const device_added& Event)
+		        [this](const audio_device_added& Event)
 		        {
-			        if (IsInput(Event.device))
 			        {
-				        InputDevices->OnAdded(Event.device);
+				        FScopeLock Lock{&AccessDevices};
+				        Devices.Add(Event.device);
+			        }
+			        if (IsInput(Event.device.direction()))
+			        {
+				        DLB_UE_LOG_DEVICE("Input", "added", Event.device.name());
+				        UpdateCurrentInputDeviceIndex();
+				        Observer.OnListOfInputDevicesChanged();
 			        }
-			        if (IsOutput(Event.device))
+			        if (IsOutput(Event.device.direction()))
 			        {
-				        OutputDevices->OnAdded(Event.device);
+				        DLB_UE_LOG_DEVICE("Output", "added", Event.device.name());
+				        UpdateCurrentOutputDeviceIndex();
+				        Observer.OnListOfOutputDevicesChanged();
 			        }
 		        })
-		    .then([this](auto&& Event) { Events->AddEvent(MoveTemp(Event)); })
-		    .on_error(DLB_HANDLE_ASYNC_EXCEPTION);
+		    .on_error(MakeHandler(__LINE__));
 
 		DeviceManagement
 		    .add_event_handler(
-		        [this](const device_removed& Event)
+		        [this](const audio_device_removed& Event)
 		        {
-			        InputDevices->OnRemoved(Event.uid);
-			        OutputDevices->OnRemoved(Event.uid);
+			        EDirection Direction{EDirection::none};
+			        {
+				        FScopeLock Lock{&AccessDevices};
+				        const auto Index = Devices.IndexOfByPredicate([UidToRemove = Event.uid](const auto& Uid)
+				                                                      { return Uid == UidToRemove; });
+				        DLB_UE_LOG_DEVICE("In/Output", "removed", Devices[Index].name());
```

Comment:
In/Output looks strange. If we can't distinguish (can't we?) between input or output, then either say Input/output or nothing at all.